### PR TITLE
feat: allow to use custom validation rules for registration

### PR DESCRIPTION
**Pull request by jozefrebjak**
This PR simplifies the validation code during registration and gives us the option to change the rules if needed without the need to extend Controller.

### CODE REVIEW

**Review by MGatner — COMMENTED** [review]
I think this would make more sense down next to `$validFields` since they influence each other. 

I am fine with this approach but I want @kenjis to give his feedback because he preferred the controller method. I do think this Config file is getting very large, but maybe a better approach would be to split out logical sections into their own Config (like `Passwords`).

**Review by kenjis — DISMISSED** [review]
LGTM!

**Comment by lonnieezell** [comment]
Should this be in Config\Validation, instead of the Auth file? That's the [user guide recommendation](https://codeigniter.com/user_guide/libraries/validation.html#saving-sets-of-validation-rules-to-the-config-file).

**Comment by jozefrebjak** [comment]
@kenjis I agree with lonnie that validation rules needs to be in Config\Validation. Then they can be called directly and we don't need to store them with Auth conf. 

Something like: `Validation.php` ?

```php
<?php

namespace CodeIgniter\Shield\Config;

use Config\Validation as BaseValidation;

class Validation extends BaseValidation
{
    /**
     * --------------------------------------------------------------------
     * Register validation rules
     * --------------------------------------------------------------------
     * Returns the rules that should be used for user register validation.
     */
    public array $register = [
        'username'         => 'required|alpha_numeric_space|min_length[3]|is_unique[users.username]',
        'email'            => 'required|valid_email|is_unique[auth_identities.secret]',
        'password'         => 'required|strong_password',
        'password_confirm' => 'required|matches[password]',
    ];
}
```

**Comment by kenjis** [comment]
Do you mean something like this?

- Add `CodeIgniter\Shield\Config\AuthValidation`
- Add `Config\AuthValidation` (extends `CodeIgniter\Shield\Config\AuthValidation`) during setup

**Comment by jozefrebjak** [comment]
Maybe something like this. Now I'm thinking. Do we need to create another config file ? It's not enough to extend base `Validation.php` like:

- Add `CodeIgniter\Shield\Config\Validation` (extends `Config\Validation`) as services now. 

and when we need to change `register` rules in App we just write own rules for `register` in `App/Config/Validation` and they will be used ? I don't know if something like this will work and if it's good practice.

**Comment by kenjis** [comment]
When `Config\Validation` exists, `config()` can't access `CodeIgniter\Shield\Config\Validation`.
Probably `setting()` also can't.

**Comment by MGatner** [comment]
`Config\Validation` is the ideal place for these rules *but* it makes it hard or impossible for developers to adjust them. This is a deficiency of the framework itself. We have two options:

1. Extend `Config\Validation` with **src/Config/Validation.php**, but this will be ignored if the developer has **app/Config/Validation.php**
2. Use `Registrar` to apply the rules to the existing `Validation` class, but then developers have no way to override them

Putting the rules in `Auth` Config at least let's developers override them.

A third option would be for our controller could check first for `Config\Validation::$registerRules` if we wanted to allow that as a source, then fall back on a default ruleset (which could live in the controller, since it wouldn't need overriding).

**Comment by MGatner** [comment]
Actually after typing that all up I like the third option the best. `getValidationRules()` simply becomes `return setting('Validation.registration') ?? [...` and nothing else had to change.

**Comment by lonnieezell** [comment]
I like the third option a lot. Keeping the logic in the controller for easy overriding there, but also checking the Validation config file should the user prefer that. That's the simple solution I knew had to be there lol. 

We would need docs around that with example of how to change, of course.

File: docs/7 - customization.md [hunk 1]
```diff
@@ -81,3 +81,51 @@ class LoginController extends ShieldLogin
     }
 }
 ```
+
+## Custom validation rules
+
+### Registration
+
+The shield has the following rules for registration:
```
Comment (MGatner) [inline docs/7 - customization.md:89]:
```suggestion
Shield has the following rules for registration:
```

File: docs/7 - customization.md [hunk 2]
```diff
@@ -81,3 +81,51 @@ class LoginController extends ShieldLogin
...
+```php
+[
+    'username'         => 'required|alpha_numeric_space|min_length[3]|is_unique[users.username]',
+    'email'            => 'required|valid_email|is_unique[auth_identities.secret]',
+    'password'         => 'required|strong_password',
+    'password_confirm' => 'required|matches[password]',
+];
+```
+
+If you need a different set of rules for registration, you can specify them in `Validation` configuration. You can use the configuration in `app\Config\Validation` and specify registration rules like:
```
Comment (MGatner) [inline docs/7 - customization.md:100]:
```suggestion
If you need a different set of rules for registration, you can specify them in your `Validation` configuration (**app\Config\Validation.php**) like:
```

File: docs/7 - customization.md [hunk 3]
```diff
@@ -81,3 +81,51 @@ class LoginController extends ShieldLogin
...
+public $registration = [
+    'username'         => 'required|alpha_numeric_space|min_length[3]|is_unique[users.username]',
+    'email'            => 'required|valid_email|is_unique[auth_identities.secret]',
+    'password'         => 'required|strong_password',
+    'password_confirm' => 'required|matches[password]',
+];
+```
+
+Or use `Registrar`. If you are not aware of the Registrar, you can read more about the class [here](https://codeigniter4.github.io/CodeIgniter4/general/configuration.html#registrars).
+
```
Comment (MGatner) [inline docs/7 - customization.md:115]:
```suggestion
For example:

```

File: docs/7 - customization.md [hunk 4]
```diff
@@ -81,3 +81,51 @@ class LoginController extends ShieldLogin
...
+```php
+/**
+ * Registers the validation.
```
Comment (MGatner) [inline docs/7 - customization.md:118]:
```suggestion
 * Registers validation rules for registration.
```

**Review by MGatner — APPROVED** [review]
This looks great! I'm glad we chewed on this for a bit, this feels right. I made some doc suggestions, take them or leave them.

**Review by MGatner — APPROVED** [review]

File: docs/7 - customization.md [hunk 5]
```diff
@@ -81,3 +81,54 @@ class LoginController extends ShieldLogin
...
+//--------------------------------------------------------------------
+public $registration = [
+    'username'         => 'required|alpha_numeric_space|min_length[3]|is_unique[users.username]',
+    'email'            => 'required|valid_email|is_unique[auth_identities.secret]',
+    'password'         => 'required|strong_password',
+    'password_confirm' => 'required|matches[password]',
+];
+```
+
+Or use `Registrar`. If you are not aware of the Registrar, you can read more about the class [here](https://codeigniter4.github.io/CodeIgniter4/general/configuration.html#registrars).
```
Comment (kenjis) [inline docs/7 - customization.md:114]: @MGatner I don't see the use case that devs use Registrar like this to their apps.

Registrar needs existing config properties. So if we use Registrar, we need to do:
1. set `Config\Validation::$registration`
2. add `Registrar` to override the property values

Comment (jozefrebjak) [reply docs/7 - customization.md:114]: I'm using it like this now. My validation rules are defined only in Registrar, so I don't need to touch Validation config at all.

Comment (kenjis) [reply docs/7 - customization.md:114]: Oh really? 
Then the `Registrar` injects properties dynamically even if there are no properties in the config class?

Dynamic properties will be deprecated.
https://wiki.php.net/rfc/deprecate_dynamic_properties

Comment (kenjis) [reply docs/7 - customization.md:114]: @jozefrebjak You are correct. Registrar can add new properties.
https://github.com/codeigniter4/CodeIgniter4/blob/fe1dbc22003fb93d1b57addd3c35fe484cf5aedb/tests/system/Config/BaseConfigTest.php#L252-L255

Comment (kenjis) [reply docs/7 - customization.md:114]: I've confirmed that will not work on PHP 8.2.
https://github.com/codeigniter4/CodeIgniter4/issues/6162

Comment (jozefrebjak) [reply docs/7 - customization.md:114]: @kenjis good to know. So we need to define it in Validation config only. What is a purpose of the Registrar then ? Just to override packages configs ? I liked how it works now.

Comment (MGatner) [reply docs/7 - customization.md:114]: To be clear, I think in this case the property should be added to **app/Config/Validation.php**; using Registrar is an second option for cases where that isn't available (like Module implementation, or Local namespace).

**Review by lonnieezell — APPROVED** [review]
I think we're all agreed this can merge?

**Comment by kenjis** [comment]
The implementation is okay.

But I hope to remove the description about Registrars in the docs.
Dynamic Property injection by Registrars is under discussion.
At least most users do not need to use Registrar in this case.

**Comment by jozefrebjak** [comment]
> The implementation is okay.
> 
> 
> 
> But I hope to remove the description about Registrars in the docs.
> 
> Dynamic Property injection by Registrars is under discussion.
> 
> At least most users do not need to use Registrar in this case.

@lonnieezell @kenjis @MGatner So, remove it ?

**Comment by MGatner** [comment]
Yes, please remove the reference to Registrars. Whether Registrars will support dynamic properties in the future is under deliberation, but I think we are close to deciding against it. It also isn't necessary here, since the same thing can be accomplished using **app/Config/Validation.php**. If a developer really needs this feature they can always allow dynamic properties on their own Config files.

**Review by kenjis — APPROVED** [review]

**Comment by kenjis** [comment]
@jozefrebjak Thank you!

**Pull request merged**